unity/Hello_World/Assets/Scripts/SpatialMappingCapture.cs

```diff
@@ -160,49 +259,40 @@ private void DetectionCallback(ObjectDetection3dSetMsg msg)
+            UnityEngine.Vector3 worldPoint = worldTransformMatrix.MultiplyPoint(p);
+            //meshPoints.Add(new PointMsg(Convert.ToDouble(worldPoint.x), Convert.ToDouble(worldPoint.y), Convert.ToDouble(worldPoint.z)));
+            meshPoints.Add(new PointMsg(Convert.ToDouble(worldPoint.x), Convert.ToDouble(worldPoint.y), Convert.ToDouble(worldPoint.z)));
 
-    public virtual void OnObservationAdded(MixedRealitySpatialAwarenessEventData<SpatialAwarenessMeshObject> eventData)
-    {
-        // Copy over the mesh data we need because we cannot access it outside of the main thread
-        Vector3[] vertices = eventData.SpatialObject.Filter.mesh.vertices;
-        int[] triangles = eventData.SpatialObject.Filter.mesh.triangles;
-        int id = eventData.SpatialObject.Id;
-        Thread tSendObject = new Thread(() => SendMeshObjectAddition(vertices,
-                                                                     triangles,
-                                                                     id));
-        tSendObject.Start();
-    }
+        }
 
-    public virtual void OnObservationUpdated(MixedRealitySpatialAwarenessEventData<SpatialAwarenessMeshObject> eventData)
-    {
-        // Copy over the mesh data we need because we cannot access it outside of the main thread
-        Vector3[] vertices = eventData.SpatialObject.Filter.mesh.vertices;
-        int[] triangles = eventData.SpatialObject.Filter.mesh.triangles;
-        int id = eventData.SpatialObject.Id;
-        Thread tSendObject = new Thread(() => SendMeshObjectAddition(vertices,
-                                                                     triangles,
-                                                                     id));
-        tSendObject.Start();
-    }
+        // Add triangle indices to the mesh triangles list
+        for (int i = 0; i < triangles.Length; i += 3)
+        {
+            meshTriangles.Add(new MeshTriangleMsg(new uint[3] {Convert.ToUInt32(triangles[i]),
+                                                               Convert.ToUInt32(triangles[i + 1]),
+                                                               Convert.ToUInt32(triangles[i + 2])
+                                                              }
+                                                 ));
+        }
 
-    public virtual void OnObservationRemoved(MixedRealitySpatialAwarenessEventData<SpatialAwarenessMeshObject> eventData)
-    {
-        int id = eventData.SpatialObject.Id;
-        Thread tSendObject = new Thread(() => SendMeshObjectRemoval(id));
-        tSendObject.Start();
+        MeshMsg shapeMsg = new MeshMsg(meshTriangles.ToArray(), meshPoints.ToArray());
+
+        // Build and return the spatial mesh message
+        SpatialMeshMsg meshMsg = new SpatialMeshMsg(id, false, shapeMsg);
+
+        return meshMsg;
     }
 
     private void SendMeshObjectAddition(Vector3[] vertices, int[] triangles, int id)
```

Comment:
We at some point need to turn a critical eye to performance, and one random thought is if this method, being thread executed, is causing some of that slow down? E.g. I don't have a feel for how often this gets called and how expensive it is. What jumps out at me is that, by threading this like we do above, we're acing in a fire-and-forget manner, meaning there there is nothing functionally stopping there being more than a managable amount of these threads doing mesh serialization.

Reply:
Yeah, that is a good point. I haven't looked at how long this takes and how often it happens, but just ball parking there's usually ~20 active meshes. Spatial awareness system is configured to update once per second (we can slow this down), so this could get called up to 20 times per second.